### ISSUE TITLE
[naming-convention] Provide a way to customize lint messages

### ISSUE THREAD

**Issue by octogonz** [typescript-eslint/typescript-eslint#2240]
**Repro**

Here's the sort of rule someone might make if they want `camelCase`, but allowing a prefix of TWO underlines instead of the hardwired `leadingUnderscore` setting:

```JS
'@typescript-eslint/naming-convention': [
    'error',
    {
        selector: 'default',
        format: [ ],

        custom: {
          regex: '^(__?)?[a-z][a-z0-9]*([A-Z][a-z0-9]*)*$',
          match: true
        }
    },
]
```

```ts
// error @typescript-eslint/naming-convention: Variable name `Oops` must
// match the RegExp: /^(__?)?[a-z][a-z0-9]*([A-Z][a-z0-9]*)*$/
const Oops: number = 123;
```

**Expected Result**

There should be some way to provide a custom message explaining the required syntax, for example:

```
The identifier ${ID} should be camelCase, possibly with an underscore prefix.
```

**Actual Result**

`@typescript-eslint/eslint-plugin` tries to generate an English sentence that describes the specification implied by its rather Byzantine options. It does not do a good job, because (1) it does not know which aspects of the spec are important to mention (e.g. double-underscores are a trivial detail), nor (2) can it infer that this RegExp implied the "camelCase" concept.

**Additional Info**

I found the new `@typescript-eslint/eslint-plugin` options to be a bit frustrating.  I still have not managed to upgrade to `3.3.0`, because I'm still struggling to craft a set of options that accurately reproduces what the old deprecated rules did. (We're trying to avoid adding hundreds of new suppressions with this upgrade.)  As another example, we permit a single underscore for cases like `checkBox1_onChanged`, but this is a very proprietary requirement.

**Suggested Solution**

Overall `@typescript-eslint/eslint-plugin` would be much more intuitive if I could simply declare what I want, rather than trying to map our needs onto someone else's concepts of `camelCase`, `leadingUnderscore`, etc.

All we really need is:

- Selectors (which is quite well designed in the current implementation, just needs `namespace`)
- A list of RegExp patterns that we will accept (plus maybe some edge cases to reject)
- An English lint message to show if none of the patterns matches

For even better lint messages, we could use (a different set of) loose RegExp's to select alternative messages based on context.

Here's a rough sketch of this idea:
```js
{
  selector: 'interface',
  // Input must match at least one of these
  allowRegexes: [
    '^_?I([A-Z][a-z0-9]+)+$'
  ],
  // Input must not match any of these
  forbidRegexes: [
  ],
  // Supplementary messages to show in specific contexts:
  alternativeFailureMessages: [
    {
      regex: '^_?I',
      match: true,
      text: 'The interface name ${ID} should use PascalCase'      
    }
  ],
  // The default message to show if none of the alternative messages is matched
  failureMessage: 'The interface name ${ID} requires an I prefix',
}
```

This is admittedly more verbose than the existing design.  But the advantages might be:

- The meaning is very explicit; no need to guess what each option does, no confusion about option precedence
- The weird extra edge cases are easy to add and fine tune
- The user experience shows professionally authored explanations rather than machine-generated text

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.3.0` |
| `@typescript-eslint/parser`        | `3.3.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `7.2.0` |
| `node`                             | `12.17.0` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
The problem that we had with the existing multi-rule solution is that each rule worked in a different way, had very little flexibility, and applied to distinct sets of things.

This lead to people asking for things like "I want this rule, but with this slight change" or "I want to check this in this circumstance", "I want properties named like this when this and like this when this", etc. It became clear that having one rule per use case wasn't really scalable. That's when a contributor showed me [this tslint rule](https://github.com/ajafff/tslint-consistent-codestyle/blob/master/docs/naming-convention.md). It took me a little bit to detangle it, but I saw the value in having a single rule to define your entire naming convention.

I thought about it for a while and figured that it was really the best option - one rule, one codebase, one consistent set of options, and infinitely customisable.

It has a tradeoff in the form of a higher barrier to entry, but I thought that could be covered in the form of many examples in the docs and sensible defaults.

So far, I _think_ I was right - most of the issues I've seen have been for more customisability, and a few for clarifications in the docs.

Though it's so hard to tell in this world. Some people just suffer in silence, some silently appreciate the rule, some chose to not use it, some raise issues to improve the system. It's so hard to figure out  when you're on the right trajectory. I usually just assume "no news is good news".

/rant

----

> no need to guess what each option does, 

What is unclear about each option? I tried to be clear with the naming.

> no confusion about option precedence

Is this unclear? I tried to be explicit about how the rule worked in the docs:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/naming-convention.md#format-options

----

I think that the request for config for multiple underscores would be of use to you #1712, as that would mean the use case you mentioned wouldn't have to use a custom regex at all.

----

As we discussed on another issue - `namespaces` should be added as a selector. It was an oversight that they weren't.

----

Definitely don't see a problem with allowing you to define a message along with a custom regex to help clarify. Regexes are rarely clear on their own, so there's definite value in allowing overrides.

Multiple custom regexes seems like a good idea as well.

```diff
 custom?:
   | {
     regex: string;
     match: boolean;
+    failureMessage?: string;
   }
+  | {
+    regex: string;
+    match: boolean;
+    failureMessage?: string;
+  }[];
```

Unlike with multiple `format `strings - where an identifier just has to match one of the array of `format`s, for an array of `custom` an identifier should match _all_ of `custom`s provided.

**Comment by octogonz** [comment]
> So far, I _think_ I was right - most of the issues I've seen have been for more customisability, and a few for clarifications in the docs.

Having worked with `naming-convention` for a while, I agree with your design choice. Although the configuration is more complex, it is much easier to specify the naming rules all together. TypeScript names can appear in all sorts of contexts, so it's simpler to only have to learn a single model for describing them.  And the "selector" concept makes it easier to avoid interactions between naming policies.

> Is this unclear? I tried to be explicit about how the rule worked in the docs:

Here's a punch list of stuff that was not obvious from the docs:
- What happens when you specify `custom.regex` and `format` together? Is it AND or OR?
- What about with `custom.match=false` (i.e. *must not* match the regex) -- does that widen or narrow the inputs accepted by `format`?
- How does `custom.regex` interact with options like `prefix`/`suffix`/etc?
- What is the default value for `leadingUnderscore`? The default value for `Options` is documented, but no defaults are given for individual fields (e.g. if I completely replace `defaultOptions` with my own structure)
- If the `filter` predicate is false, do other option blocks get evaluated and applied instead (i.e. potentially matching some other selector)?  Or is lint validation simply skipped if the `filter` predicate is false?
- What happens when a `modifiers` constraint is not relevant? For example `memberLike` includes `method` and `enumMember`, but the `protected` modifier is impossible for `enumMember`.  Is `enumMember` rejected, or is `protected`  ignored?
- Similarly what happens with `types` applied to `memberLike` which includes some items that do support `types`? I see that the docs say `Allowed types: none`, but no error was reported when I tried it (if I remember right).

**Comment by octogonz** [comment]
> Unlike with multiple `format `strings - where an identifier just has to match one of the array of `format`s, for an array of `custom` an identifier should match _all_ of `custom`s provided.

Is your reasoning that the regex language itself easily supports `alternative1|alternative2|alternative3`, whereas it can be difficult to make a pattern that satisfies `rule1` AND `rule2` AND `rule3`?

That's technically true, although in my case [I had to use a script macro](https://github.com/microsoft/rushstack/blob/e1428a0ad6b576f55b0f6982f4bfed40cacac180/stack/eslint-config/index.js#L169) to make the OR blocks more readable, like this:
```js
  filter: {
    regex: [
      // This is a special exception for naming patterns that use an underscore to separate two camel-cased
      // parts.  Example:  "checkBox1_onChanged" or "_checkBox1_onChanged"
      '^_?[a-z][a-z0-9]*([A-Z][a-z]?[a-z0-9]*)*_[a-z][a-z0-9]*([A-Z][a-z]?[a-z0-9]*)*$'
    ]
      .map((x) => `(${x})`)
      .join('|'),
    match: false
  }
```
This works, but one of our repos has a lint test that generates a snapshot of the entire lint configuration (since is complex and pulls in rule configurations from libraries). The snapshot is not very readable when it contains expanded macros.  So if there was some way for `naming-convention` to support this directly, it might be valuable.

**Comment by bradzacher** [comment]
> Unlike with multiple format strings - where an identifier just has to match one of the array of formats, for an array of custom an identifier should match all of customs provided.

Probably should have explained it, yeah.

There was a few reasons behind this limitation:

***First*** - if you allow OR for `custom` then you can get into a weird state if you have both `match: false` and `match: true`.

For example:
```json5
[
  {
    regex: /^I/,
    match: true,
  },
  {
    regex: /^I/,
    match: false,
  }
]
```

This config would essentially do nothing, because all identifiers match these buckets.
In more complex cases, it would be pretty hard to validate that you aren't putting yourself into this case.

***Second*** - it was because of the new `failureMessage` option.

If you have 10 `custom` items, and an identifier fails them all, what would the error message look like? Concatenate them all together into one mega lint message?

```
variable name `Foo` must:
- not match the RegExp: /<some regex>/. Custom Message 1
- not match the RegExp: /<some regex>/. Custom Message 2
- not match the RegExp: /<some regex>/. Custom Message 3
- not match the RegExp: /<some regex>/. Custom Message 4
- not match the RegExp: /<some regex>/. Custom Message 5
- not match the RegExp: /<some regex>/. Custom Message 6
- not match the RegExp: /<some regex>/. Custom Message 7
- not match the RegExp: /<some regex>/. Custom Message 8
- not match the RegExp: /<some regex>/. Custom Message 9
- not match the RegExp: /<some regex>/. Custom Message 10
```

That would be pretty scary to read as a user!

Do you pick just one of the messages? If you do - that would be pretty misleading to the user, as they wouldn't get a proper picture of what's allowed, and a small typo that causes the name to not match the 6th regex would be super confusing to resolve.

***Finally*** - it would add so much variation to the rule that it would become really confusing to understand:

```js
{
  selector: 'variable',
  format: ['UPPER_CASE', 'camelCase', 'PascalCase'],
  custom: [
    {
      regex: /<regex 1>/,
      match: true,
    },
    {
      regex: /<regex 2>/,
      match: false,
    },
    {
      regex: /<regex 3>/,
      match: true,
    },
  ],
}
```

This config would mean "a variable must be in UPPER_CASE OR camelCase OR PascalCase, AND it must either match <regex 1> OR not match <regex 2> OR match <regex 3>".

I don't know about you, but reasoning about which identifiers that actually allows seems really confusing to me.

**Comment by sonsina** [comment]
"As we discussed on another issue - namespaces should be added as a selector. It was an oversight that they weren't."

@bradzacher is there any intention of adding something about this?

Thanks

**Comment by nathanchase** [comment]
Yes, I really want to be able to be clear about how this rule functions to any team member:
```json
{
  "rules": {
    "@typescript-eslint/naming-convention": [
      "error",
      {
        "selector": "interface",
        "format": ["PascalCase"],
        "custom": {
          "regex": "^(?!I[A-Z])[A-Z][a-zA-Z0-9]+$",
          "match": true,
          "message": "Interfaces should be in PascalCase, and should not be prefixed with 'I'."
        }
      }
    ]
  }
}

```

**Comment by JoshuaKGoldberg** [comment]
Per #8792, we're feature freezing this rule. It's reached the point of complexity where we're having a hard time justifying working on it. 

If you'd like to see this option happen, I'd encourage you to add it to a community project such as [eslint.style](https://eslint.style). Cheers! 💙

**Comment by karlhorky** [comment]
I also mentioned this just now over in my question to the ESLint Stylistic maintainers about potentially migrating this rule to ESLint Stylistic:

- https://github.com/typescript-eslint/typescript-eslint/issues/8792#issuecomment-2025344673
